docs/src/proposals/append-vec-storage.md

```diff
@@ -0,0 +1,231 @@
+---
+title: Reduce the size of AppendVec storage in AccountsDB
+---
+
+## Problems
+The storage format, StoredAccountMeta, of each entry in the AppendVec currently
+has 129 bytes overhead + the size of account data for each account.
+
+StoredAccountMeta currently has the following format:
+
+```
+pub struct StoredAccountMeta<'a> {
+    pub meta: &'a StoredMeta {
+        pub write_version: StoredMetaWriteVersion, // 8 bytes
+        pub pubkey: Pubkey,  // 32 bytes, encoded from a 44-byte string.
+        pub data_len: u64,   //  8 bytes
+    },
+    pub account_meta: &'a AccountMeta {
+        pub lamports: u64,     //  8 bytes
+        pub owner: Pubkey,     // 32 bytes, encoded from a 44-byte string
+        pub executable: bool,  //  1 byte
+        pub rent_epoch: Epoch, //  8 bytes
+    },
+    pub data: &'a [u8],     //  N bytes
+    pub offset: usize,      //  0 bytes in storage, 8 bytes in memory
+    pub stored_size: usize, //  0 bytes in storage, 8 bytes in memory
+    pub hash: &'a Hash,     // 32 bytes
+}
+
+```
+
+## Proposed Solution
+The new AppendVec storage format uses ~80 bytes for each account in average,
+and it's also more compression friendly for account data.
+
+Specifically:
+
+* Rent-exempt account w/o shared owner uses 80 bytes each.
+* Rent-exempt account w/ shared owner uses 40 + 32 / N bytes each,
+  where N is the number of accounts sharing its owner.
+* Non-rent exempt account uses an extra 8 bytes for rent epoch.
+
+The new AppendVec file is organized into 6 different blocks: header,
+account metas block, account pubkeys block, owners block, account data block,
+and footer.
+
+```
++---------------------------------------------------+
+| header                | 40 bytes                  |
++---------------------------------------------------+
+| account metas block   | 16 bytes per account      |
++---------------------------------------------------+
+| account pubkeys block | 32 bytes per account      |
++---------------------------------------------------+
+| owners block          | 32 bytes per unique owner |
++---------------------------------------------------+
+| account data block    | N + optional 8 bytes      |
+|                       | for rent epoch            |
++---------------------------------------------------+
+| footer                | 40 bytes                  |
++---------------------------------------------------+
+```
+
+### Better Compression for Account Data
+All account-data is put together into one block to make them easier
+to compress, compared to having them stored interleaving with the pubkeys.
+
+### Per-Account Hash is replaced by AppendVec Block Hash
+The per-account hash is removed.  Instead, each block in the new AppendVec file
+has one 32-bytes hash, and the hash for the entire AppendVec is also included
+in the footer.
```

Comment:
why a hash in the footer for this append vec? This is not deterministic per machine.

Reply:
The footer includes the AppendVec hash + a magic number.  If the ending part of one AppendVec files does not match the predefined magic number, then we know for sure that the file has been truncated or incompleted.

Reply:
Removed the footer.  Move the AppendVec hash to the header instead.